lib/snowflake_db/connection.rb

```diff
@@ -0,0 +1,56 @@
+require 'odbc_utf8'
+require 'sequel'
+module SnowflakeDB
+  class Connection
+    attr_reader :config, :unload_target, :max_file_size
+
+    PG_INCLUDE_KEYS = %w(host dbname port user password)
+    MAX_FILE_SIZE = (5*1024*1024*1024).freeze
+
+    def initialize(config)
+      @config = config
+      @statement_timeout = config['statement_timeout']
+      @unload_target = config['unload_target']
+      @max_file_size = config['max_file_size'] || MAX_FILE_SIZE
+    end
+
+    def reconnect_on_failure(&block)
+      begin
+        return yield
+      rescue Sequel::DatabaseError => e
+        raise unless connection_expired_error?(e)
+        connection.reset
+        return yield
+      end
+    end
+
+    def connection
+      @connection ||= connect!
+    end
+
+    def fetch_all_hash(query)
+      # execute a query and return the entire result as an array of hashes keyed by column names
+      result = []
+      reconnect_on_failure do
+        connection.fetch(query) do |row|
+          result << Hash[row.map { |k, v| [k.to_s, v] }]
```

Comment:
again, since we have active_support, might as well just use `result << row.stringify_keys`